File: cpp/interactive-base-game.cpp
#include <iostream>
#include <vector>
#include <random>
#include <bitset>
#include <array>
#include <algorithm>
#include <iomanip>

class InteractiveBaseGame {
private:
    std::bitset<12> board;
    std::array<std::vector<int>, 13> subsetCache;
    std::array<int, 2> scores;
    std::mt19937 gen;
    std::uniform_int_distribution<> dist;

    void initSubsetCache() {
        for (int i = 2; i <= 12; ++i) {
            for (int j = 1; j < i; ++j) {
                if ((1 << (j - 1)) & ((1 << i) - 1)) {
                    subsetCache[i].push_back(j);
                }
            }
        }
    }

    int rollTwoDice() {
        return dist(gen) + dist(gen);
    }

    void displayBoard() const {
        std::cout << "+---+---+---+---+---+---+---+---+---+---+---+---+\n";
        std::cout << "| ";
        for (int i = 11; i >= 0; --i) {
            std::cout << (board[i] ? "O" : "X") << " | ";
        }
        std::cout << "\n+---+---+---+---+---+---+---+---+---+---+---+---+\n";
        std::cout << "  12  11  10   9   8   7   6   5   4   3   2   1  \n";
    }

public:
    InteractiveBaseGame(unsigned seed = std::random_device{}())
        : board(0xFFF), scores{0, 0}, gen(seed), dist(1, 6)
    {
        initSubsetCache();
    }

    void play() {
        int rounds;
        std::cout << "Enter the number of rounds to play: ";
        std::cin >> rounds;

        for (int round = 0; round < rounds; ++round) {
            std::cout << "Round " << round + 1 << " begins:" << std::endl;
            for (int player = 0; player < 2; ++player) {
                board = 0xFFF;
                int turnScore = playerTurn(player);
                scores[player] += turnScore;
                std::cout << "Player " << (player == 0 ? "A" : "B") << " round score: " << turnScore << std::endl;
                std::cout << "Current total: Player A = " << scores[0] << ", Player B = " << scores[1] << std::endl << std::endl;
            }
        }

        std::cout << "Game over. Final scores: Player A = " << scores[0] 
                  << ", Player B = " << scores[1] << std::endl;
        std::cout << "Winner: Player " << (scores[0] < scores[1] ? "A" : "B") << std::endl;
    }

    int playerTurn(int player) {
        std::cout << "Player " << (player == 0 ? "A" : "B") << "'s turn begins. Initial board:" << std::endl;
        displayBoard();
        
        while (true) {
            std::cout << "Press Enter to roll the dice...";
            std::cin.ignore();
            std::cin.get();

            int roll = rollTwoDice();
            std::cout << "You rolled: " << roll << std::endl;

            if (board[roll - 1]) {
                std::cout << "Do you want to turn off bit " << roll << "? (y/n): ";
                char choice;
                std::cin >> choice;
                if (choice == 'y' || choice == 'Y') {
                    board[roll - 1] = 0;
                    std::cout << "Turned off bit " << roll << std::endl;
                } else {
                    std::cout << "Searching for subset..." << std::endl;
                    bool subsetFound = false;
                    for (int num : subsetCache[roll]) {
                        if (board[num - 1]) {
                            std::cout << "Do you want to turn off bit " << num << "? (y/n): ";
                            std::cin >> choice;
                            if (choice == 'y' || choice == 'Y') {
                                board[num - 1] = 0;
                                subsetFound = true;
                                std::cout << "Turned off bit " << num << std::endl;
                                break;
                            }
                        }
                    }
                    if (!subsetFound) {
                        int score = board.count();
                        std::cout << "No valid moves. Turn ends with score: " << score << std::endl;
                        std::cout << "Final board state:" << std::endl;
                        displayBoard();
                        return score;
                    }
                }
            } else {
                std::cout << "Bit " << roll << " is already off. Searching for subset..." << std::endl;
                bool subsetFound = false;
                for (int num : subsetCache[roll]) {
                    if (board[num - 1]) {
                        std::cout << "Do you want to turn off bit " << num << "? (y/n): ";
                        char choice;
                        std::cin >> choice;
                        if (choice == 'y' || choice == 'Y') {
                            board[num - 1] = 0;
                            subsetFound = true;
                            std::cout << "Turned off bit " << num << std::endl;
                            break;
                        }
                    }
                }
                if (!subsetFound) {
                    int score = board.count();
                    std::cout << "No valid moves. Turn ends with score: " << score << std::endl;
                    std::cout << "Final board state:" << std::endl;
                    displayBoard();
                    return score;
                }
            }

            std::cout << "Current board:" << std::endl;
            displayBoard();

            if (board.none()) {
                std::cout << "All bits turned off! Perfect score of 0." << std::endl;
                return 0;
            }
        }
    }
};

File: rust/interactive-base-game.rs
use rand::Rng;
use std::io::{self, Write};

pub struct InteractiveBaseGame {
    board: u16,
    subset_cache: Vec<Vec<u16>>,
    scores: [u32; 2],
    rng: rand::rngs::ThreadRng,
}

impl InteractiveBaseGame {
    pub fn new() -> Self {
        let mut game = InteractiveBaseGame {
            board: 0xFFF,
            subset_cache: vec![vec![]; 13],
            scores: [0; 2],
            rng: rand::thread_rng(),
        };
        game.init_subset_cache();
        game
    }

    fn init_subset_cache(&mut self) {
        for i in 2..=12 {
            for j in 1..i {
                if (1 << (j - 1)) & ((1 << i) - 1) != 0 {
                    self.subset_cache[i].push(j as u16);
                }
            }
        }
    }

    fn roll_two_dice(&mut self) -> u16 {
        self.rng.gen_range(1..=6) + self.rng.gen_range(1..=6)
    }

    fn display_board(&self) {
        println!("+---+---+---+---+---+---+---+---+---+---+---+---+");
        print!("| ");
        for i in (0..12).rev() {
            print!("{} | ", if self.board & (1 << i) != 0 { "O" } else { "X" });
        }
        println!("\n+---+---+---+---+---+---+---+---+---+---+---+---+");
        println!("  12  11  10   9   8   7   6   5   4   3   2   1  ");
    }

    pub fn play(&mut self) {
        println!("Enter the number of rounds to play:");
        let mut input = String::new();
        io::stdin().read_line(&mut input).expect("Failed to read line");
        let rounds: i32 = input.trim().parse().expect("Please enter a number");

        for round in 0..rounds {
            println!("Round {} begins:", round + 1);
            for player in 0..2 {
                self.board = 0xFFF;
                let turn_score = self.player_turn(player);
                self.scores[player] += turn_score as u32;
                println!("Player {} round score: {}", if player == 0 { "A" } else { "B" }, turn_score);
                println!("Current total: Player A = {}, Player B = {}\n", self.scores[0], self.scores[1]);
            }
        }

        println!("Game over. Final scores: Player A = {}, Player B = {}", self.scores[0], self.scores[1]);
        println!("Winner: Player {}", if self.scores[0] < self.scores[1] { "A" } else { "B" });
    }

    fn player_turn(&mut self, player: usize) -> u16 {
        println!("Player {}'s turn begins. Initial board:", if player == 0 { "A" } else { "B" });
        self.display_board();
        
        loop {
            println!("Press Enter to roll the dice...");
            io::stdin().read_line(&mut String::new()).expect("Failed to read line");

            let roll = self.roll_two_dice();
            println!("You rolled: {}", roll);

            if self.board & (1 << (roll - 1)) != 0 {
                println!("Do you want to turn off bit {}? (y/n):", roll);
                let mut choice = String::new();
                io::stdin().read_line(&mut choice).expect("Failed to read line");
                if choice.trim().to_lowercase() == "y" {
                    self.board &= !(1 << (roll - 1));
                    println!("Turned off bit {}", roll);
                } else {
                    println!("Searching for subset...");
                    let mut subset_found = false;
                    for &num in &self.subset_cache[roll as usize] {
                        if self.board & (1 << (num - 1)) != 0 {
                            println!("Do you want to turn off bit {}? (y/n):", num);
                            let mut sub_choice = String::new();
                            io::stdin().read_line(&mut sub_choice).expect("Failed to read line");
                            if sub_choice.trim().to_lowercase() == "y" {
                                self.board &= !(1 << (num - 1));
                                subset_found = true;
                                println!("Turned off bit {}", num);
                                break;
                            }
                        }
                    }
                    if !subset_found {
                        let score = self.board.count_ones();
                        println!("No valid moves. Turn ends with score: {}", score);
                        println!("Final board state:");
                        self.display_board();
                        return score as u16;
                    }
                }
            } else {
                println!("Bit {} is already off. Searching for subset...", roll);
                let mut subset_found = false;
                for &num in &self.subset_cache[roll as usize] {
                    if self.board & (1 << (num - 1)) != 0 {
                        println!("Do you want to turn off bit {}? (y/n):", num);
                        let mut choice = String::new();
                        io::stdin().read_line(&mut choice).expect("Failed to read line");
                        if choice.trim().to_lowercase() == "y" {
                            self.board &= !(1 << (num - 1));
                            subset_found = true;
                            println!("Turned off bit {}", num);
                            break;
                        }
                    }
                }
                if !subset_found {
                    let score = self.board.count_ones();
                    println!("No valid moves. Turn ends with score: {}", score);
                    println!("Final board state:");
                    self.display_board();
                    return score as u16;
                }
            }

            println!("Current board:");
            self.display_board();

            if self.board == 0 {
                println!("All bits turned off! Perfect score of 0.");
                return 0;
            }
        }
    }
}